tianshou/policy/modelfree/redq.py

```diff
@@ -55,72 +47,86 @@ class REDQPolicy(DDPGPolicy):
+            action_space=action_space,
+            tau=tau,
+            gamma=gamma,
+            exploration_noise=exploration_noise,
+            estimation_step=estimation_step,
+            action_scaling=action_scaling,
+            action_bound_method=action_bound_method,
+            observation_space=observation_space,
+            lr_scheduler=lr_scheduler,
         )
-        self.actor, self.actor_optim = actor, actor_optim
-        self.critics, self.critics_old = critics, deepcopy(critics)
-        self.critics_old.eval()
-        self.critics_optim = critics_optim
-        assert 0 < subset_size <= ensemble_size, "Invalid choice of ensemble size or subset size."
         self.ensemble_size = ensemble_size
         self.subset_size = subset_size
 
-        self._is_auto_alpha = False
-        self._alpha: float | torch.Tensor
-        if isinstance(alpha, tuple):
-            self._is_auto_alpha = True
-            self._target_entropy, self._log_alpha, self._alpha_optim = alpha
-            assert alpha[1].shape == torch.Size([1])
-            assert alpha[1].requires_grad
-            self._alpha = self._log_alpha.detach().exp()
-        else:
-            self._alpha = alpha
-
-        if target_mode in ("min", "mean"):
-            self.target_mode = target_mode
-        else:
-            raise ValueError("Unsupported mode of Q target computing.")
-
+        self.target_mode = target_mode
         self.critic_gradient_step = 0
         self.actor_delay = actor_delay
-        self._deterministic_eval = deterministic_eval
+        self.deterministic_eval = deterministic_eval
         self.__eps = np.finfo(np.float32).eps.item()
 
-    def train(self, mode: bool = True) -> "REDQPolicy":
-        self.training = mode
-        self.actor.train(mode)
-        self.critics.train(mode)
-        return self
+        # TODO: reduce duplication with SACPolicy
+        self.alpha: float | torch.Tensor
+        self._is_auto_alpha = not isinstance(alpha, float)
+        if self._is_auto_alpha:
+            # TODO: why doesn't mypy understand that this must be a tuple?
+            alpha = cast(tuple[float, torch.Tensor, torch.optim.Optimizer], alpha)
+            if alpha[1].shape != torch.Size([1]):
+                raise ValueError(
+                    f"Expected log_alpha to have shape torch.Size([1]), "
+                    f"but got {alpha[1].shape} instead.",
+                )
+            if not alpha[1].requires_grad:
+                raise ValueError("Expected log_alpha to require gradient, but it doesn't.")
+
+            self.target_entropy, self.log_alpha, self.alpha_optim = alpha
+            self.alpha = self.log_alpha.detach().exp()
+        else:
+            # TODO: make mypy undestand this, or switch to something like pyright...
+            alpha = cast(float, alpha)
+            self.alpha = alpha
+
+    @property
+    def is_auto_alpha(self) -> bool:
+        return self._is_auto_alpha
```

Comment:
why not just replace `_is_auto_alpha` to `is_auto_alpha`?

Reply:
Because the user doesn't pass `is_auto_aplha`, it is inferred from the `alpha` input (if it's a tuple or a constant). So the user code should never be able to mutate it, hence the property